### ISSUE TITLE
A100 8 GPUs extremely slow compared to a single A100

### ISSUE THREAD

**Issue by KeremTurgutlu** [google/jax#15962]
### Description

TL;DR When I run a t5x script using a A100-8 GPU machine it is much slower compared to running the same script on a single A100 machine.

There are many available configurations for running language model training using the t5x library which is based on jax/flax. To reproduce this issue one can take any of the training scripts run it on a single A100 and then do the same on a A100 8 GPU machine. I also cached the tfds dataset and read it from local disk to minimize any overhead and to just compare performance when switching from 1 GPU to 8 GPU single node training. The model I use doesn't require model parallelism, e.g. you can pick up any model size up to (m)T5 XL (3.7B) to reproduce.

Here are the stats I got during my training runs:
 
```
# On A100 8 GPU machine - extremely slow / slow compile warning
seqs_per_second_per_core=0.156312

# On single A100 machine / runs fine 
seqs_per_second_per_core=~6
```

```
********************************
[Compiling module pjit_train_step] Very slow compile? If you want to file a bug, run with envvar XLA_FLAGS=--xla_dump_to=/tmp/foo and attach the results.
********************************
```

I am not able to attach the file as zst format is not supported and zip compressed folder is too large.

### What jax/jaxlib version are you using?

jax==0.4.9 jaxlib==0.4.9+cuda12.cudnn88

### Which accelerator(s) are you using?

GPU

### Additional system info

Python 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] :: Anaconda, Inc. on linux

### NVIDIA GPU info

```
Thu May 11 09:01:07 2023       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 530.30.02              Driver Version: 530.30.02    CUDA Version: 12.1     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                  Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf            Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A100-SXM4-80GB           On | 00000000:05:00.0 Off |                    0 |
| N/A   30C    P0               77W / 400W|  61847MiB / 81920MiB |    100%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-SXM4-80GB           On | 00000000:06:00.0 Off |                    0 |
| N/A   29C    P0              127W / 400W|  61835MiB / 81920MiB |    100%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   2  NVIDIA A100-SXM4-80GB           On | 00000000:07:00.0 Off |                    0 |
| N/A   35C    P0              152W / 400W|  61835MiB / 81920MiB |    100%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   3  NVIDIA A100-SXM4-80GB           On | 00000000:08:00.0 Off |                    0 |
| N/A   30C    P0               79W / 400W|  61835MiB / 81920MiB |    100%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   4  NVIDIA A100-SXM4-80GB           On | 00000000:09:00.0 Off |                    0 |
| N/A   34C    P0               81W / 400W|  61835MiB / 81920MiB |    100%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   5  NVIDIA A100-SXM4-80GB           On | 00000000:0A:00.0 Off |                    0 |
| N/A   30C    P0               86W / 400W|  61835MiB / 81920MiB |    100%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   6  NVIDIA A100-SXM4-80GB           On | 00000000:0B:00.0 Off |                    0 |
| N/A   34C    P0               84W / 400W|  61835MiB / 81920MiB |    100%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   7  NVIDIA A100-SXM4-80GB           On | 00000000:0C:00.0 Off |                    0 |
| N/A   33C    P0               78W / 400W|  61835MiB / 81920MiB |    100%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A    271492      C   python3                                   61776MiB |
|    1   N/A  N/A    271492      C   python3                                   61764MiB |
|    2   N/A  N/A    271492      C   python3                                   61764MiB |
|    3   N/A  N/A    271492      C   python3                                   61764MiB |
|    4   N/A  N/A    271492      C   python3                                   61764MiB |
|    5   N/A  N/A    271492      C   python3                                   61764MiB |
|    6   N/A  N/A    271492      C   python3                                   61764MiB |
|    7   N/A  N/A    271492      C   python3                                   61764MiB |
+---------------------------------------------------------------------------------------+
```

**Comment by KeremTurgutlu** [comment]
@nouiz @yhtang Thanks for replying quickly. Let me give more details below by trying to reproduce the issue on 2 GPU A100 machine (I couldn't get a 8 A100 this time I will try again with it once available):

TL;DR It worked as intended in a 2 A100 GPU - more details below. 

### Software Stack

Running [umt5 xl model](https://github.com/google-research/t5x/blob/main/t5x/examples/scalable_t5/umt5/pretrain_xl.gin).

Model implementation uses jax pjit and according to [README](https://github.com/google-research/t5x/blob/main/t5x/examples/scalable_t5/README.md) should support TPU (most efficient-but pricey) and GPU clusters.

I am not using SLURM yet, right now for initial benchmarking I am using a script to do single node training with 8 GPUs where distribution strategy is only data parallelism. Later, I plan to try a larger model on single node and also move the training to a GPU cluster based on the price-performance compared to TPUs. 

Training script:

```bash
#!/bin/bash
# set train hyperparams
unset LD_PRELOAD

# set environment variables
PROJECT_DIR=${HOME}"/<DIRECTORY_WITH_GIN_FILES>"
T5X_DIR=${HOME}"/t5x"  # directory where the t5x is cloned.
TFDS_DATA_DIR="<NOT_SO_IMPORTANT_LOCAL_CACHED_DATA_IS_USED>"
export PYTHONPATH=${PROJECT_DIR}

export XLA_PYTHON_CLIENT_PREALLOCATE=true
export XLA_PYTHON_CLIENT_MEM_FRACTION=0.9 # default is less around 61/80 GB

python3 ${T5X_DIR}/t5x/train.py \
  --gin_search_paths=${PROJECT_DIR} \
  --gin_file="xl_umt5_pretrain.gin" \
  --tfds_data_dir=${TFDS_DATA_DIR}
```

Content of `xl_umt5_pretrain.gin`:

```python
from __gin__ import dynamic_registration
import tasks

import __main__ as train_script
from t5x import utils
from t5x import partitioning
from t5x import trainer
import seqio

include "t5x/examples/scalable_t5/umt5/pretrain_xl.gin"


MODEL_DIR = "<MODEL_DIR_TO_SAVE>"
MIXTURE_OR_TASK_NAME = "<SEQIO_MIXTURE_NAME>"
TRAIN_EVAL_MIXTURE_OR_TASK_NAME = "<SEQIO_MIXTURE_NAME>"
LOSS_NORMALIZING_FACTOR = None
# LOSS_NORMALIZING_FACTOR = "NUM_REAL_TARGET_TOKENS"
DROPOUT_RATE = 0.0
TRAIN_STEPS = 1_011_200  # 1_000_000 pretrained steps + 11_200 pre-finetuning
TASK_FEATURE_LENGTHS = {"inputs": 512, "targets": 512}
RANDOM_SEED = 42
# BATCH_SIZE = 1024
# BATCH_SIZE = 16
BATCH_SIZE = 128 # 16 x 8 = 128 with 8 A100 GPUs
SHUFFLE_TRAIN_EXAMPLES = False # shuffles files for tfds data source
USE_CACHED_TASKS = True


# ------------------- Training specification overrides --------------------------
train_script.train:
  eval_period = 2000
  stats_period = 20
  eval_steps = 20
  random_seed = %RANDOM_SEED # (used for dropout) but just to be sure in general
  train_dataset_cfg = @train/utils.DatasetConfig()
  # train_eval_dataset_cfg = @train_eval/utils.DatasetConfig()
  train_eval_dataset_cfg = None # gradient accumulation not working?
  infer_eval_dataset_cfg = None
  run_eval_before_training = False # OOM with microbatches when True
  partitioner = @partitioning.PjitPartitioner()

partitioning.PjitPartitioner:
  model_parallel_submesh = (1,1,1,1)
  num_partitions = None

trainer.Trainer:
  num_microbatches = 0 # 1024 / 32 = 32 (bs 32 tpu v3-8)

utils.create_learning_rate_scheduler:
  factors = 'rsqrt_decay'
  step_offset = 960_000 # step = 1_000_000 - 960_000, 1 / sqrt(40_000) = 5e-3

utils.DatasetConfig:
  seed = 42 # read_config.shuffle_seed in the case of TfdsDataSource

utils.SaveCheckpointConfig:
  period = 2000
  save_dataset = True
  # keep = 5 # default is None which keeps all checkpoints

utils.RestoreCheckpointConfig:
  mode = 'specific'
  dtype= 'bfloat16'
  path = 'gs://t5-data/pretrained_models/t5x/umt5_xl/checkpoint_1000000'
  assignment_map = None
  strict = True
  dtype = None
```

### Machine Specs

1000 GB RAM | 128 vCPUs | 4000GB SSD

[lscpu.info.txt](https://github.com/google/jax/files/11468415/lscpu.info.txt)
[cpu.info.txt](https://github.com/google/jax/files/11468416/cpu.info.txt)



Base Ubuntu 20.04

Dependency Installation Steps:

1) CUDA:

```
wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/cuda-keyring_1.0-1_all.deb
sudo dpkg -i cuda-keyring_1.0-1_all.deb
sudo apt-get update
sudo apt-get -y install cuda
```

2) Miniconda:

```
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
...# install miniconda
conda create -n myenv python=3.9
```

3) Verify jax with GPUs:

```
python -c "import jax; print(jax.device_put(jax.numpy.ones(1), device=jax.devices('gpu')[0]).device())"
python -c "import jax; print(jax.devices())"
```

```
gpu:0
[StreamExecutorGpuDevice(id=0, process_index=0, slice_index=0), StreamExecutorGpuDevice(id=1, process_index=0, slice_index=0)]
```

4) T5X:

```
git clone https://github.com/google-research/t5x.git
cd t5x
pip install -e .
# also install dependencies
pip install t5
```

5) Download cached dataset to machine.


6) Run single node training scripts

### Running with 1 device visible with `CUDA_VISIBLE_DEVICES=0`

```
PROJECT_DIR=
T5X_DIR=
TFDS_DATA_DIR=
export PYTHONPATH=${PROJECT_DIR}

# Setting XLA flags
export XLA_FLAGS="--xla_gpu_simplify_all_fp_conversions \
--xla_gpu_all_reduce_combine_threshold_bytes=136314880 ${XLA_FLAGS}\
--xla_dump_to=xla_run_dump"

export CUDA_VISIBLE_DEVICES=0
export XLA_PYTHON_CLIENT_PREALLOCATE=true
export XLA_PYTHON_CLIENT_MEM_FRACTION=0.9

python3 ${T5X_DIR}/t5x/train.py \
  --gin_search_paths=${PROJECT_DIR} \
  --gin_file="xl_umt5_pretrain.gin" \
  --tfds_data_dir=${TFDS_DATA_DIR}
```

[single_gpu_train.log](https://github.com/google/jax/files/11468566/single_gpu_train.log)

**Note:** Interestingly I got OOM with a
[single_gpu_train_bs8_oom.log](https://github.com/google/jax/files/11468564/single_gpu_train_bs8_oom.log)
 batch size = 8 - in a single machine I was able to train with a batch size of 16.

Adding training logs:
[single_gpu_train.log](https://github.com/google/jax/files/11468578/single_gpu_train.log)


### Running with all 2 devices (assuming data parallelism via JAX pjit)

Adding training logs:

[two_gpus_train.log](https://github.com/google/jax/files/11468580/two_gpus_train.log)


Performance per core / gpu is similar but need to try again with 8 GPUs.


1 gpu training: `timing/seqs_per_second_per_core=5.6435`

2 gpu training: `timing/seqs_per_second_per_core=4.85872`

**Comment by hawkinsp** [comment]
Is this still a problem? We've done lots of work on LLM training in JAX, and we expect it to be screamingly fast these days on A100 GPUs.
